### DISCUSSION TITLE
component testing with react testing library

### DISCUSSION THREAD

**Discussion by ufelgen** [neuefische/questions#442] [Web] [answered]
Hey everyone, 

I am having some trouble with component testing. 
1. I am trying to tell the test what specific sub-page it's currently on. I read some things online and tried with Memory Router

```jsx
import { render, screen } from "@testing-library/react";
import Footer from "./Footer";
import { MemoryRouter } from "react-router-dom";

test("renders more info button when on declension page", async () => {
  render(
    <MemoryRouter initialEntries={["/declension/[id]"]}>
      <Footer />
    </MemoryRouter>
  );
  const infoButton = screen.getByLabelText("show more info");
  expect(infoButton).toBeInTheDocument();
});
```

but I can't get it to work. Is that something that is possible with react testing library? If yes, how?

2. I am also trying to test if a popup component is rendered after button click. I tried many things but none of them work unfortunately. I tried mocking the popup component but I think I'm missing a step somewhere. 

```jsx
jest.mock("../PopupMenu/PopupMenu", () => () => (
  <div data-testid="popup">Hello World</div>
));

test("renders popup menu upon popup menu button click", async () => {
  const user = userEvent.setup();
  const handlePopupClick = jest.fn();

  render(
    <VocabCard word={word} popup={false} onPopupClick={handlePopupClick} />
  );

  const popupButton = screen.getByRole("button", { name: "open popup menu" });

  await user.click(popupButton);
  const mockPopupMenu = screen.getByTestId("popup");

  expect(mockPopupMenu).toBeVisible();
});
```

the button sets the "popup" state to the corresponding id, which is the condition for rendering the component. How can I test correctly for this?

**Thanks for your help!**

**Comment by actyralf** [comment]
The problem in your first question could be that you need to initialize MemoryRouter with a specific id and not with the placeholder `[id]`.
Can you provide the error messages you get and a link to your PR / branch?

**Reply by ufelgen** [reply]
Thank you! I opened a PR: https://github.com/ufelgen/capstone-project/pull/34

Error messages: 
<img width="855" alt="Bildschirmfoto 2023-01-09 um 10 32 56" src="https://user-images.githubusercontent.com/115288914/211278313-b4a8aa6d-8330-4e06-96e4-75359a34a30d.png">

<img width="857" alt="Bildschirmfoto 2023-01-09 um 10 33 22" src="https://user-images.githubusercontent.com/115288914/211278337-a9502800-4869-44b4-aaf0-30bbbf73699d.png">

<img width="743" alt="Bildschirmfoto 2023-01-09 um 10 33 41" src="https://user-images.githubusercontent.com/115288914/211278345-c39eddf1-53ab-4777-8a47-1047d8f69b9c.png">

**Comment by pixelass** [comment]
I updated your question with code blocks with the correct syntax highlighting.
Have a look at the suggested format for code blocks:

- https://github.com/neuefische/questions/wiki/How-to-ask-good-questions#react
- https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

❤️‍🔥

**Reply by ufelgen** [reply]
Thank you, that helps! sorry I missed that.

**Comment by actyralf** [comment, marked as answer]
Ah - you are mixing up React Router & Next Router.
React Router is not used in your app, so you can remove the `MemoryRouter` stuff.
To mock `useRouter` you can install this package: `npm i --save-dev next-router-mock`

Then you should add these lines to your test:

```js
import mockRouter from 'next-router-mock';
jest.mock('next/router', () => require('next-router-mock'));
jest.mock('next/dist/client/router', () => require('next-router-mock'));
```

In your tests you can set the current path BEFORE rendering like this:

```js
mockRouter.setCurrentUrl("/declension/123");
```
I tested this and it passes the url to your footer component. The test still fails - but I think this is a problem in the component. I'm not sure what you want to do, but you check if `pathname === "/declension/[id]"`. It will never have this value because `[id]` is a placeholder.
If you have a concrete url like `/declension/123` which renders the footer correctly, the test should pass.

**Reply by ufelgen** [reply]
Thanks!! I got it green with: 

````js
import { render, screen } from "@testing-library/react";
import Footer from "./Footer";
import mockRouter from "next-router-mock";
import "@testing-library/jest-dom";

jest.mock("next/router", () => require("next-router-mock"));
jest.mock("next/dist/client/router", () => require("next-router-mock"));

test("renders more info button when on declension page", async () => {
  mockRouter.setCurrentUrl("/declension/[id]");
  render(<Footer />);
  const infoButton = screen.getByLabelText("show more info");
  expect(infoButton).toBeInTheDocument();
});
````

**Comment by pixelass** [comment]
You are using react-router on top of next. This will cause issues. Please have a look at mocking routes for nextjs.
Also don't use `querySelector` but instead use the internal helpers from testing library.


- https://dev.to/peterlidee/how-to-mock-next-router-with-jest-3p6b
- https://testing-library.com/docs/queries/bytestid/

**Reply by ufelgen** [reply]
Thanks, I was able to fix this issue

**Comment by ufelgen** [comment]
Thank you for your help! any ideas concerning the second problem?

**Reply by pixelass** [reply]
Take a look at `PopupMenuButton`. You are looking for a button with a certain `name`

```js
const popupButton = screen.getByRole("button", { name: "open popup menu" });
```

which expects a DOM element similar to this:

```html
<button name="open popup menu">Click me</button>
```

First off: `name` should be ONE word e.g. `button:menuOpen`. Since it is not common to give a button a `name`, you could try to select the button with `screen.getByText` or another valid selector method. 

Also, it looks like setting the open state relies on outside logic: `word.id === popup`. That ,means that you have to build a "dummy component" to test the behavior or decide to skip the "integration test" and focus on the component test. Here you could test if a click is really tracked. 

```js
expect(handlePopupClick).toHaveBeenCalled()
```

**Reply by ufelgen** [reply]
yes, it definitely relies on outside logic, specifically the popup useState that lives in the _app.js. The test works with .toHaveBeenCalled(), but that doesn't test for my popup component being rendered or not. If there isn't a way to do this, I will go with checking if the function has been called. thank you!

**Reply by pixelass** [reply]
as mentioned, you could build the necessary logic in a "dummy component" and render this in a test.

(This is an untested example. I'm not sure if this is how your component works)

This tries to isolate the use-case to confirm that the integration in a statefuil component works as expected.

```jsx

function MyTestComponent() {
  const [popup, setPopup] =- useState(false)

  return (
    <VocabCard
      word="Hello"
      popup={popup}
      onPopupClick={() => {
        setPopup("Hello")
      }} />
  )
}

```

**Reply by ufelgen** [reply]
awesome!! yes this worked with a little fiddling: 

```js
function MyTestComponent() {
  const [popup, setPopup] = useState(false);

  return (
    <>
      <VocabCard
        word={word}
        popup={popup}
        onPopupClick={() => {
          setPopup(word.id);
        }}
      />
      <p data-testid="popup">{popup}</p>
    </>
  );
}

test("sets 'popup' state to word-id upon button click so that popup will open", async () => {
  const user = userEvent.setup();
  render(<MyTestComponent />);
  const popup = screen.getByTestId("popup");

  const popupButton = screen.getByRole("button", { name: "open popup menu" });

  await user.click(popupButton);

  expect(popup).toHaveTextContent(77);
});

```

Thanks!

**Reply by pixelass** [reply]
Actually… I think this test is not valid, since the popup is now always visible. Since I don't know the entire logic of your app, I can't make a suggestion. You would have to replicate the entire popup logic. Right now you are basically just testing if the click works (sane as testing if the function has been called). What you want to do is to check if the popup is open (very similar to your first attempt with `beInTheDocument`)

**Reply by ufelgen** [reply]
I think I see what you mean. Except I'm testing if the function is called with the correct id... and it doesn't work without the user click

**Reply by pixelass** [reply]
That is true but the `word` is also defined in the same level/scope (in your test file or folder) therefore it is obvious that the test passes. You are just testing if the id is rendered as the text, wich is obvious for this test case. 

Your test successfully proves that when clicking the button a function is called. What you do in this function is not relevant since you could also check if the function `hasBeenCalled`.

Same scope: 

https://github.com/ufelgen/capstone-project/pull/34/files#diff-6859a085e7c06a649e63334a8c127aefa6dd80a96b175e1dab5a227c65cfc238R8-R41

![Screenshot 2023-01-09 at 15 49 00](https://user-images.githubusercontent.com/1148334/211335989-23c11060-52d0-47c7-920a-0564d32da5ae.png)
